src/reader.rs

```diff
@@ -174,18 +174,29 @@ where
         unsafe { rest.set_len(max) };
 
         while self.buffer.len() < target_size {
-            let n = ready!(poll_reader(Pin::new(&mut self.reader), cx, &mut rest[..]))?;
-            if n == 0 {
-                if self.buffer.is_empty() {
-                    return Poll::Ready(Ok(FillResult::EOF));
-                } else {
-                    return Poll::Ready(Err(io::Error::from(io::ErrorKind::BrokenPipe)));
+            match poll_reader(Pin::new(&mut self.reader), cx, &mut rest[..]) {
+                Poll::Ready(result) => {
+                    let n = result?;
```

Comment:
I wonder if we should also recover the capacity if we hit an `Err` here? It's less likely to matter than the (common) `Pending` case, but might as well just handle it too.